[FLINK-21187] Provide exception history for root causes

XComp opened the pull request:
## What is the purpose of the change

This PR handles the initial exception history proposal collecting only the root causes. This is part of a major effort of providing a exception history to the Flink web UI ([FLINK-6042](https://issues.apache.org/jira/browse/FLINK-6042)).


## Brief change log

* updated `AccessExecutionGraph` interface to return `ErrorInfo` instead of a stringified version of the failure.
* FailureHandlingResult get a new member identifying the `ExecutionVertex` that failed.
* Added two maps collecting the error infos for each occurring failure root cause.

## Verifying this change

* `ExecutionFailureHandlerTest` and `FailureHandlingResultTest` were updated accordingly.
* `DefaultSchedulerTest#testExceptionHistoryWithRestartableFailure` was introduced to cover the exception history collection.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java:
```diff
@@ -870,6 +874,78 @@ public void allocationIsCanceledWhenVertexIsFailedOrCanceled() throws Exception
+        ErrorInfo globalFailure = actualExceptionHistory.get(1);
+        Throwable actualException =
+                globalFailure.getException().deserializeError(ClassLoader.getSystemClassLoader());
+        assertThat(actualException, org.hamcrest.core.IsInstanceOf.instanceOf(JobException.class));
+        assertThat(actualException.getCause(), is(failingException));
```
One could use `FlinkMatchers.containsCause`.

XComp replied:
Ah, thanks for the hint. I didn't know about `FlinkMatchers`, yet.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java:
```diff
@@ -870,6 +874,78 @@ public void allocationIsCanceledWhenVertexIsFailedOrCanceled() throws Exception
+            JobID jobId,
+            ExecutionAttemptID executionAttemptID,
+            Throwable exception) {
+        long start = System.currentTimeMillis();
```
`System.currentTimeMillis()` can be susceptible to clock resets. Hence, it can cause some test instabilities. Unfortunately, `Execution` also uses this to determine the state timestamp.

XComp replied:
Correct me if I'm wrong here: Another way to measure the time would be `System.nanoTime()` since it ensures monotonicity. But using this in the test would require `Execution` to use `System.nanoTime()` as well which would not work as `System.nanoTime` cannot be used to compare timestamps between different JVMs [[1]](https://www.javaadvent.com/2019/12/measuring-time-from-java-to-kernel-and-back.html)?!

tillrohrmann replied:
Yes, we are actually interested in a timestamp and not some measure of passed time. But for testing purposes a manually controlled `Clock` can help to avoid instabilities cause by `System.currentTimeMillis`. The production code could then still use `System.currentTimeMillis`.

XComp replied:
Thanks for clarification.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:
```diff
@@ -641,6 +648,33 @@ public void cancel() {
+        globalTaskFailures.put(
+                executionAttemptIds, new ErrorInfo(failure, System.currentTimeMillis()));
+
+        log.debug("Archive global {}: {}", failure, failure.getMessage());
```
How would this log line look like? Wouldn't `failure.getMessage` be duplicated? Maybe also say "Archive global failure...".

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:
```diff
@@ -164,6 +168,8 @@
     private final Map<OperatorID, OperatorCoordinatorHolder> coordinatorMap;
 
     private final ComponentMainThreadExecutor mainThreadExecutor;
+    private final Map<ExecutionAttemptID, ErrorInfo> localTaskFailures = new HashMap<>();
+    private final Map<Set<ExecutionAttemptID>, ErrorInfo> globalTaskFailures = new HashMap<>();
```
For what do we need the maps here? Wouldn't a single `List<ErrorInfo>` be good enough here?

XComp replied:
You're right. That was still some leftover of my initial work considering multiple exceptions per failure cause. I'm gonna refactor it to use a List here in order to be more consistent.

tillrohrmann reviewed (changes requested):
Thanks for creating this PR @XComp. I had a couple of questions and comments. Please take a look.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResult.java:
```diff
@@ -100,17 +116,22 @@ public long getRestartDelayMS() {
+     *
+     * @return The {@code ExecutionVertexID} or {@code null} if it's a global failure.
+     */
+    public ExecutionVertexID getExecutionVertexIdOfFailedTask() {
```
I switched to returning an `Optional` to make this more explicit.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:
```diff
@@ -641,6 +648,33 @@ public void cancel() {
+    }
+
+    protected void archiveFromFailureHandlingResult(FailureHandlingResult failureHandlingResult) {
+        if (failureHandlingResult.getExecutionVertexIdOfFailedTask() == null) {
```
I removed this check. This if was a relict from a previous code version.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java:
```diff
@@ -267,9 +267,16 @@ private void restartTasksWithDelay(final FailureHandlingResult failureHandlingRe
+                                                        archiveFromFailureHandlingResult(
+                                                                failureHandlingResult),
+                                                getMainThreadExecutor())
+                                        .thenRunAsync(
+                                                restartTasks(
+                                                        executionVertexVersions, globalRecovery),
+                                                getMainThreadExecutor())),
```
That's a good question. I did it when experimenting for convenience and didn't reiterate over it again. 👍 I refactored it now.

XComp commented:
Thanks @tillrohrmann for reviewing the PR. I addressed your changes. I removed code that would become only necessary when adding the support for the exception history collecting concurrent failures to be more consistent.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/AccessExecution.java:
```diff
@@ -64,9 +66,11 @@
      *
-     * @return failure exception as a string, or {@code "(null)"}
+     * @return an {@code Optional} of {@link ErrorInfo} containing the {@code Throwable} wrapped in
+     *     a {@link SerializedThrowable} and the time it was registered if an error occurred. If no
```
nit: That the throwable is wrapped in a `SerializedThrowable` should be an implementation detail of the `ErrorInfo`. I think it is not necessary to mention it here.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ArchivedExecution.java:
```diff
@@ -121,8 +125,8 @@ public AllocationID getAssignedAllocationID() {
-    public String getFailureCauseAsString() {
-        return failureCause;
+    public Optional<ErrorInfo> getFailureInfo() {
+        return failureInfo == null ? Optional.empty() : Optional.of(failureInfo);
```
```suggestion
        return Optional.ofNullable(failureInfo);
```

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java:
```diff
@@ -317,13 +316,11 @@ public TaskManagerLocation getAssignedResourceLocation() {
+    public Optional<ErrorInfo> getFailureInfo() {
+        return failureCause == null
+                ? Optional.empty()
+                : Optional.of(new ErrorInfo(failureCause, getStateTimestamp(FAILED)));
```
Does it make sense to change `failureCause` to be of type `ErrorInfo`? If yes, then we could save recreating this object multiple times.

XComp replied:
Yes, it does. I actually had it like that in a first version of the code but changed it for the sake of consistency after realizing that it's not Serializable (which was a problem in `ArchivedExecution`). I will revert it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResult.java:
```diff
@@ -100,17 +126,26 @@ public long getRestartDelayMS() {
+     *     it's a global failure.
+     */
+    public Optional<ExecutionVertexID> getExecutionVertexIdOfFailedTask() {
+        return failingExecutionVertexId == null
+                ? Optional.empty()
+                : Optional.of(failingExecutionVertexId);
```
```suggestion
        return Optional.ofNullable(failingExecutionVertexId);
```

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:
```diff
@@ -522,6 +527,7 @@ protected ComponentMainThreadExecutor getMainThreadExecutor() {
     protected void failJob(Throwable cause) {
         incrementVersionsOfAllVertices();
         executionGraph.failJob(cause);
+        getTerminationFuture().thenRun(() -> archiveGlobalFailure(cause));
```
I think we should check here that we are indeed in the `FAILED` state. What can also happen is that the user cancels the job in between.

XComp replied:
Here, I don't understand fully: You're referring to the case where a failure happens, the user cancels the job while the failure handling is done and the `failJob` method might be called while being in a `CANCELED`/`CANCELLING` state?For this case, I would still think that we should archive the exception because the users intervention happened after the exception happened.

tillrohrmann replied:
I mean the case that a failure happens, the job goes into the `FAILING` state and tries to cancel the tasks and now the user cancels the job because it takes too long for him. Then the job will go into the `CANCELING` state which will result to the `CANCELED` state once all tasks have terminated.

I think you are right that we should still record the failure cause.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:
```diff
@@ -641,6 +647,32 @@ public void cancel() {
     }
 
+    protected void archiveGlobalFailure(Throwable failure) {
+        taskFailureHistory.add(new ErrorInfo(failure, System.currentTimeMillis()));
```
I would suggest to take the `ExecutionGraph.getStatusTimestamp`. That way the timestamp will be consistent.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ArchivedExecutionGraphTest.java:
```diff
@@ -371,9 +375,12 @@ private static void compareExecution(
+                runtimeExecution.getFailureInfo().map(ErrorInfo::getExceptionAsString),
+                is(archivedExecution.getFailureInfo().map(ErrorInfo::getExceptionAsString)));
+        assertThat(
+                runtimeExecution.getFailureInfo().map(ErrorInfo::getTimestamp),
+                is(archivedExecution.getFailureInfo().map(ErrorInfo::getExceptionAsString)));
```
Is this correct? Why does CI passes with this assertion if it is wrong?

XComp replied:
Oh man, good catch. `ArchiveExecutionGraphTest` triggered a global failure which only set the `failureInfo` on the `ExecutionGraph` level. I switched to `SchedulerBase.updateTaskExecutionState` which also sets the `failureInfo` on an `Execution` level

tillrohrmann reviewed (approved):
Thanks for updating this PR @XComp. I had a few minor comments. After resolving them +1 for merging.

XComp commented:
@pnowojski @AHeise @rkhachatryan Could one of you guys have a look at the `Task`-related change I did in [211a68e](https://github.com/apache/flink/pull/14798/commits/211a68e9532b0bf10581388b395e474123ec6ad5)?

rkhachatryan reviewed (changes requested):
I took a look at `[FLINK-21187][runtime] Adds invariant for FAILED state always having a cause` commit.
I'm concerned about updating the `failureCause` before actually transitioning (see below).

I also understand there is some code path where the cause is null. Is it correct? What was this place?
I think it should be changed in the same commit.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -1022,18 +1021,38 @@ private boolean transitionState(ExecutionState currentState, ExecutionState newS
+    private boolean transitionToFailedStateAndSetFailureCause(
+            ExecutionState currentState, Throwable cause) {
+        Preconditions.checkNotNull(cause, "No cause is given when transitioning to FAILED state.");
+        failureCause = cause;
+        return transitionState(currentState, ExecutionState.FAILED, cause);
```
That's actually a good point. I will correct that.

XComp commented:
@rkhachatryan thanks for looking into it. There was no production code path I ran into that let me start this investigation. I ran into test issues after adding the invariant check to `TaskExecutionState`. That's where I started to check the code to see whether I could rely on the invariant to be true all the time. The only place where the `failureCause` is not set even though it's transitioning into `FAILED` state is [in the fall back case](https://github.com/apache/flink/pull/14798/commits/af0bf68b965c9adaab61f42a99ef3278d27aed32#diff-b84174e55cb1999d99ad60cdeded7be20ff4978472bfc785c5a77b6270f47b56R856).

XComp commented:
The test failures reported by Azure seem to be related to [FLINK-21277](https://issues.apache.org/jira/browse/FLINK-21277). I'm going to rebase and squash the related commits together after @rkhachatryan gave his final ok

rkhachatryan reviewed (approved):
Thanks for updating the PR and clarifying @XComp.
LGTM.

XComp commented:
@tillrohrmann I addressed all comments, squashed the changes and rebased the branch.

tillrohrmann commented:
Thanks for the update. I'll merge this PR once AZP has completed.

XComp commented:
There seems to be some test instability in the fine-grained source data now with the invariant being added. The [Azure build attached to this PR](https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=13005&view=logs&j=a57e0635-3fad-5b08-57c7-a4142d7d6fa9&t=5360d54c-8d94-5d85-304e-a89267eb785a) succeeded. [Another Azure build on the same commit](https://dev.azure.com/mapohl/flink/_build/results?buildId=221&view=logs&j=cc649950-03e9-5fae-8326-2f1ad744b536&t=51cab6ca-669f-5dc0-221d-1e4f7dc4fc85&l=5837) failed in `TaskExecutorSlotLifetimeTest`. The test failure is caused by the newly introduced Precondition. To me, that looks like an issue in the fine-grained resource management code. I'd prefer to create a Jira issue covering this. I still think that the invariant makes sense and we shouldn't relax it.

tillrohrmann commented:
Ok, then let's first fix this instability before merging this PR @XComp. Ping once the instability has been fixed.

XComp commented:
The test instability is caused by the change adding the requirement that `failureCause` need to be set if the `executionState` is `FAILED`. The failing test runs a task but doesn't wait for the task to be finished. The `TaskExecutor` triggers the cancellation of the `Task` through [Task.failExternally](https://github.com/XComp/flink/blob/354eb574c98bbc5fdaedabc9f3ca7e4acfaa3746/flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java#L1129) passing in a `FlinkException: Disconnect from JobManager responsible for [...]`. This will trigger the `Task` to transition into `FAILED`. A context switch happens after the state is set but before the `failureCause` is set. Then, the `Task` finishes in the main thread, cleans up and calls [notifyFinalState](https://github.com/XComp/flink/blob/354eb574c98bbc5fdaedabc9f3ca7e4acfaa3746/flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java#L904) as part of this which initializes a `TaskExecutionState` and runs into the `IllegalStateException` since the `failureCause` is still not set.

The solution would be to do the state change and `failureCause` atomically. But I'm not sure whether this would be a performance problem. We would have to switch from [compareAndSet](https://github.com/XComp/flink/blob/354eb574c98bbc5fdaedabc9f3ca7e4acfaa3746/flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java#L1064) to a normal lock, I guess. Alternatively, we could remove the invariant again and handle the `null` in the exception history. @tillrohrmann what are your thoughts on that?

XComp commented:
I looked through the code (supported by @AHeise): The race condition really only kicks in when cancelling/failing the task because that's when the `failureCause` becomes relevant. So, instead of synchronizing the [state transition](https://github.com/XComp/flink/blob/5781449f38c1e36c1a2952518f9e30761d915f04/flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java#L1052) we could add synchronize blocks for the [cancellation of a task](https://github.com/XComp/flink/blob/5781449f38c1e36c1a2952518f9e30761d915f04/flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java#L1130) and while [handling failure handling during the normal Invokable execution](https://github.com/XComp/flink/blob/5781449f38c1e36c1a2952518f9e30761d915f04/flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java#L828). The synchronization will only cover the state transition and setting the `failureCause`. Cancelling the corresponding task would be moved out of the synchronization block.

XComp commented:
I created [FLINK-21376](https://issues.apache.org/jira/browse/FLINK-21376) to cover the `failureCause`/`executionState` synchronization. The code changes of FLINK-21187 will handle cases where no `failureCause` is set for now.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java:
```diff
@@ -153,7 +152,7 @@
     private LogicalSlot assignedResource;
 
-    private Throwable failureCause; // once assigned, never changes
+    private Optional<ErrorInfo> failureCause = Optional.empty(); // once assigned, never changes
```
The comment seems to be wrong now.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java:
```diff
@@ -1133,7 +1128,10 @@ private void processFail(
 
         // success (in a manner of speaking)
-        this.failureCause = t;
+        if (t != null) {
+            // we only set the failureCause if an error is passed (see FLINK-21376)
+            this.failureCause = Optional.of(new ErrorInfo(t, getStateTimestamp(FAILED)));
+        }
```
Can't we say that if `t == null`, then we create an unknown failure cause? `new FlinkException("Unknown cause for Execution failure. This might be caused by FLINK-21376")`?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResult.java:
```diff
@@ -38,6 +42,12 @@
+     */
+    @Nullable private final ExecutionVertexID failingExecutionVertexId;
+
     /** Reason why the failure is not recoverable. */
```
The JavaDoc seems wrong. If I am not mistaken, then we will also set this field if the failure is recoverable, right?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:
```diff
@@ -635,6 +641,33 @@ public void cancel() {
+                        .map(this::getExecutionVertex)
+                        .map(ExecutionVertex::getCurrentExecutionAttempt);
+
+        executionOptional.ifPresent(
+                execution ->
+                        execution
+                                .getFailureInfo()
+                                .ifPresent(
+                                        failureInfo -> {
+                                            taskFailureHistory.add(failureInfo);
+                                            log.debug(
+                                                    "Archive local failure causing attempt {} to fail: {}",
+                                                    execution.getAttemptId(),
+                                                    failureInfo.getExceptionAsString());
+                                        }));
```
Why do we have to ask the execution for the failure cause? If we are only interested in the root cause, doesn't `failureHandlingResult` contain all the required information?

XComp replied:
Because of the timestamp in case of local failure: The `Execution` fails but does not necessarily trigger the `ExecutionGraph` to switch to fail. Hence, no timestamp is record for this kind of failure.

XComp replied:
But looking into that again, I realize that there is, indeed, the `failureInfo` in `ExecutionGraph` that is set during [handleTaskFailure](https://github.com/XComp/flink/blob/8e732bfb2bddc38ec7422f482dcda4be3d296408/flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java#L218) and [handleGlobalFailure](https://github.com/XComp/flink/blob/8e732bfb2bddc38ec7422f482dcda4be3d296408/flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java#L236). This is used to forward the current root cause of the `ExecutionGraph` to the Web UI. I could have used that. But the drawback is that it's using `System.currentMillis`.

Instead, I'm gonna work on removing the `ExecutionGraph.failureInfo` (and related methods) as part of FLINK-21188. It becomes redundant after we enable to exception history since it's just the exception history's last entry.

tillrohrmann replied:
So theoretically we would have to add the timestamp to the `FailureHandlingResult` or the `ErrorInfo` if we didn't want to look up the `Execution` again, right?

XComp replied:
Yes, that would be possible if would would like to relax the constraint that the exception history does not provide the timestamp the `Execution` was marked as failed but an earlier timestamp. But to be fair, we do that already for the fallback of a global failover in `archiveFromFailureHandlingResult`.

tillrohrmann replied:
Ok, I think we don't have to change it right now. Maybe we do this sometime in the future.

XComp replied:
I added this remark to FLINK-21190

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:
```diff
@@ -635,6 +641,33 @@ public void cancel() {
+    }
+
+    protected void archiveFromFailureHandlingResult(FailureHandlingResult failureHandlingResult) {
+        Optional<Execution> executionOptional =
+                failureHandlingResult
+                        .getExecutionVertexIdOfFailedTask()
+                        .map(this::getExecutionVertex)
+                        .map(ExecutionVertex::getCurrentExecutionAttempt);
```
I think `archiveFromFailureHandlingResult` can also be called when handling a global failover. In this case `executionOptional` would be empty and as a consequence we won't record a failure cause.

XComp replied:
You're right. I remember looking into that. Back then, I mistook the task as only focusing on "task failures". That was a misunderstanding. I added the test for global failover and the test.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultSchedulerTest.java:
```diff
@@ -870,6 +871,87 @@ public void allocationIsCanceledWhenVertexIsFailedOrCanceled() throws Exception
     }
 
+    @Test
+    public void testExceptionHistoryWithRestartableFailure() {
```
I think we also need a test for global failover.

tillrohrmann reviewed (changes requested):
Thanks for updating this PR @XComp. I've given it another pass and stumbled across some questions/potential problems where I wasn't sure whether I understood everything correctly. Please take a look.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:
```diff
@@ -635,6 +641,41 @@ public void cancel() {
         return executionGraph.getTerminationFuture().thenApply(FunctionUtils.nullFn());
     }
 
+    protected void archiveGlobalFailure(Throwable failure) {
+        archiveGlobalFailure(failure, executionGraph.getStatusTimestamp(JobStatus.FAILED));
+    }
+
+    protected void archiveGlobalFailure(Throwable failure, long timestamp) {
+        taskFailureHistory.add(new ErrorInfo(failure, timestamp));
+        log.debug("Archive global failure.", failure);
+    }
+
+    protected void archiveFromFailureHandlingResult(FailureHandlingResult failureHandlingResult) {
```
Shall we make these methods `final` since we don't want them to be overriden.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:
```diff
@@ -635,6 +641,41 @@ public void cancel() {
+        } else {
+            // fallback in case of a global fail over - no failed state is set and, therefore, no
+            // timestamp was taken
+            archiveGlobalFailure(failureHandlingResult.getError(), System.currentTimeMillis());
```
I think `getError` can return `null`. This needs to be handled properly.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java:
```diff
@@ -635,6 +641,41 @@ public void cancel() {
+                        .map(this::getExecutionVertex)
+                        .map(ExecutionVertex::getCurrentExecutionAttempt);
+
+        if (executionOptional.isPresent()) {
```
Maybe it is a bit clearer to use `failureHandlingResult.isGlobalFailure()` because at the moment the contract is that `executionVertexIdOfFailedTask == null` iff it's a global failure.

XComp replied:
I'm hesitant to change that: IMHO, it makes the code harder to read and easier to misbehave. The current version of the code is straight forward: Use the `FAILED` timestamp provided the corresponding `Execution` if we have the corresponding `ExecutionAttemptId`. Additionally, your change request might cause unwanted behavior if we decide to introduce global failures due to some local failure. I consider the fact that the global failure does not have a causing `Execution` an implementation detail which we don't have to expose here. @tillrohrmann Is that reasonable?

tillrohrmann replied:
Fair enough. Then let's not change it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/flip1/FailureHandlingResult.java:
```diff
@@ -100,17 +131,24 @@ public long getRestartDelayMS() {
      *
      * @return reason why the restarting cannot be conducted
      */
     public Throwable getError() {
```
`@Nullable` is missing. One way to find nullability problems is to turn on IntelliJ's inspections. There are couple of those for null checks. This helps me quite a lot.

tillrohrmann reviewed (approved):
Thanks for updating the PR @XComp. I think the solution looks good. Well done. I had a few last minor comments. Once they are resolved ping me to merge this PR.

XComp commented:
I addressed all changes, squashed and rebased the branch.
